Improve build times by caching layers, when possible

Ocramius opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

By using `ghcr.io/laminas/laminas-continuous-integration-action:build-cache` as a registry-based
docker image layer cache, we can avoid re-building any intermediate docker image layers for which
an exact SHA256 match exists.

This will dramatically speed up builds that happen to change little within the `Dockerfile` (or
any related sources), and it will reduce any network I/O failures caused by upstream package manager
hiccups.

Note that random users cannot just push to the cache: active caching can only happen when operating
within the boundaries of the `laminas/*` organization (direct repository pushes, operations performed
by maintainers).

To prevent stale images due to caching, we do a full re-build when the base image changes: it's an acceptable
cost, for something that changes less frequently, while we still retain caching advantages
when multiple pushes are performed close to each other.

Ocramius committed:
Improve build times by caching layers, when possible

By using `ghcr.io/laminas/laminas-continuous-integration-action:build-cache` as a registry-based
docker image layer cache, we can avoid re-building any intermediate docker image layers for which
an exact SHA256 match exists.

This will dramatically speed up builds that happen to change little within the `Dockerfile` (or
any related sources), and it will reduce any network I/O failures caused by upstream package manager
hiccups.

Note that random users cannot just push to the cache: active caching can only happen when operating
within the boundaries of the `laminas/*` organization (direct repository pushes, operations performed
by maintainers).

Ocramius committed:
Always attempt to pull a newer base image when building docker images

In a previous commit, we introduced `cache-from` and `cache-to`, which lead
to re-using docker image layers as much as possible. This is fine, but it also
means that, on the long term, we may run into stale base layers, where dependencies
do not get updated when the system shifts dramatically.

To prevent that, we do a full re-build when the base image changes: it's an acceptable
cost, for something that changes less frequently, while we still retain caching advantages
when multiple pushes are performed close to each other.

Ocramius commented:
Hmm, I'm unsure how to test this one: it seems to me that the `Dockerfile` is only used:

```yml
on:
  release:
    types: [published]
```

@weierophinney how do we verify this change? :thinking:

Ocramius commented:
In fact, I think we should test the `Dockerfile` build at every change: not just on release :thinking: 

The idea would be to have `push: false` when outside the context of a release.

weierophinney commented:
> In fact, I think we should test the `Dockerfile` build at every change: not just on release

We can add a workflow for that - it would be identical to the build-and-push-containers workflow, but with `push:false`, and operating on pull request instead of release publication. Can you add that to this PR?

Ocramius commented:
> Can you add that to this PR?

Yeah, didn't figure out how to do it without completely replicating the workflow file (which I'd like to avoid)

weierophinney commented:
Add the pull_request event to the workflow.

For the `push` value, you can use `${{ github.event_name == "release" }}`.

weierophinney committed:
feat: build on pull request as well as release

This change:

- adds "pull_request" as an event that will trigger the workflow.
- marks the "docker login" step as conditional, only to run if the "container_username" secret is present.
- alters the original "Build and push" step such that it only runs on a release
- adds two new "build and push" steps
  - one runs on pull_request if a container_username secret is present, and caches layers
  - one runs on pull_request if a container_username secret is NOT present, and DOES NOT cache layers

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
@Ocramius figured out how to make this happen.

Interestingly... it will be better in the future for TSC members to submit PRs as branches pushed directly to the repo, as otherwise repository secrets are not injected in the workflow, and we thus have no credentials for pushing to the registry.... and thus no ability to cache layers. (Release will _always_ cache layers, however.)

Pushing a change to the `README.md` to add this note about contribution, and then can merge.

weierophinney reviewed (approved):
:ship:

weierophinney committed:
docs: document need for TSC to submit patches as branches to the canonical repo

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
